File: content/problems/1930-unique-length-3-palindromic-subsequences/solutions/outer-letter-first-last/Solution.kt
class Solution {
    fun countPalindromicSubsequence(s:String):Int{
        val first=IntArray(26){
            -1
        };
        val last=IntArray(26);
        for(i in s.indices){
            val value=s[i]-'a';
            if(first[value]==-1)first[value]=i;
            last[value]=i
        };
        var answer=0;
        for(value in 0 until 26)if(first[value]+1<last[value]){
            val middle=BooleanArray(26);
            for(i in first[value]+1 until last[value])middle[s[i]-'a']=true;
            answer+=middle.count{
                it
            }
        };
        return answer
    }
}

File: content/problems/1930-unique-length-3-palindromic-subsequences/solutions/outer-letter-first-last/solution.cpp
class Solution {
    public: int countPalindromicSubsequence(string s){
        int first[26],last[26]={
        };
        fill(first,first+26,-1);
        for(int i=0;i<(int)s.size();++i){
            int value=s[i]-'a';
            if(first[value]<0)first[value]=i;
            last[value]=i;
        }
        int answer=0;
        for(int value=0;value<26;++value)if(first[value]+1<last[value]){
            bool middle[26]={
            };
            for(int i=first[value]+1;i<last[value];++i)middle[s[i]-'a']=true;
            answer+=count(middle,middle+26,true);
        }
        return answer;
    }
};
